raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -192,28 +202,45 @@ public KafkaRaftClient(NetworkChannel channel,
         this.connections = new ConnectionCache(channel, bootstrapServers,
             retryBackoffMs, requestTimeoutMs, logContext);
         this.unsentAppends = new ArrayBlockingQueue<>(10);
+        this.pendingAppends = new TreeMap<>();
```

Comment:
I was looking at the lifecycle of this collection. Seems like we are missing logic to clean it after becoming a follower?

Also, since we are maintaining this set anyway for the purpose of updating metrics, shall we go ahead and change the callback semantics as well so that the append callback is only fired after the entry is committed? Or should we do this separately?

Reply:
Re: cleaning, good one, I will fix.

Re: callback, yeah that's what I was asking on slack about as well about the future returned to state-machine (so that the sm itself would not need to maintain this logic anymore). I'm happy to make this change in a follow-up PR if we agree it is better to change the sm interface semantics.

Reply:
Yeah, i'm still not 100% sure about it, but I'm leaning in that direction. My only concern was that we have to hold onto requests longer than we would otherwise. Also, from the writer's perspective, a failure to get a response does not mean that the entry wasn't appended, but I guess as long as we are relying on idempotent operations, then retrying is ok.